lib/class-wp-block-supports.php

```diff
@@ -0,0 +1,454 @@
+		$default = (array) $default;
+		if ( count( $path ) !== count( $default ) ) {
+			return false;
+		}
+		$partial_path = array();
+		foreach ( $path as $i => $subkey ) {
+			$partial_path[] = $subkey;
+			$result         = gutenberg_experimental_get(
+				$block_supports,
+				$partial_path,
+				$default[ $i ]
+			);
+			if ( ! $result ) {
+				break;
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * TODO.
+	 *
+	 * @param  string $input      HTML input.
+	 * @param  array  $attributes Array of HTML attributes to insert at the
+	 *                            root element.
+	 * @return string             HTML with class(es) inserted.
+	 */
+	private function inject_attributes( $input, $attributes ) {
+		$output = $input;
+
+		foreach ( $attributes as $attribute_name => $attribute_value ) {
+			$close_token        = '>';
+			$attribute_token    = "$attribute_name=\"";
+			$close_position     = strpos( $output, $close_token );
+			$attribute_position = strpos( $output, $attribute_token );
+
+			// Unexpected. Bail.
+			if ( false === $close_position ) {
+				return $input;
+			}
+
+			// If the first HTML element in the string does not contain our
+			// attribute...
+			if ( ! $attribute_position || $close_position < $attribute_position ) {
+				// then inject a new one at the end of the tag.
+				$output = substr_replace(
+					$output,
+					" $attribute_name=\"$attribute_value\">",
+					$close_position,
+					strlen( $close_token )
+				);
+				continue;
+			}
+			// Otherwise, overwrite the opening of the attribute in order to
+			// inject our value.
+			$output = substr_replace(
+				$output,
+				"$attribute_name=\"$attribute_value ",
+				$attribute_position,
+				strlen( $attribute_token )
+			);
+		}
+
+		return $output;
+	}
+
+	/**
+	 * TODO.
+	 */
+	private function register_attributes() {
+		wp_register_style( 'wp-block-supports', false );
+
+		$block_registry         = WP_Block_Type_Registry::get_instance();
+		$registered_block_types = $block_registry->get_all_registered();
+		foreach ( $registered_block_types as $block_type ) {
+			if ( ! property_exists( $block_type, 'supports' ) ) {
+				continue;
+			}
+			if ( ! $block_type->attributes ) {
+				$block_type->attributes = array();
+			}
+			foreach ( $this->config as $feature_name => $feature_config ) {
+				if (
+					! $this->get_block_support(
+						$block_type->supports,
+						$feature_name,
+						$feature_config['default']
+					)
+				) {
+					continue;
+				}
+				foreach ( $feature_config['attributes'] as $attribute_name => $attribute_schema ) {
+					if ( ! array_key_exists( $attribute_name, $block_type->attributes ) ) {
+						$block_type->attributes[ $attribute_name ] = $attribute_schema;
+					}
+				}
+			}
+		}
+	}
+}
+
+/**
+ * Generates a string of classes by applying to the current block being
+ * rendered all of the features that the block supports.
+ *
+ * @return string String of HTML classes.
+ */
+function gutenberg_block_supports_classes() {
```

Comment:
I think we should try to get this as close as possible to `useBlockProps` so what about something  like `gutenberg_get_block_props` or `gutenberg_get_block_wrapper_attributes`, something generic like that.

Reply:
Also, should it takes extra props as arguments and merge theme like useBlockProps does.

Reply:
So would it look like this?

```php
// latest-posts/index.php
if ( isset( $attributes ... ) $class .= ' has-dates';
if ( isset( $attributes ... ) $class .= ' has-author';

return sprintf(
  '<ul %s>%s</ul>',
  gutenberg_get_block_wrapper_attributes( array( 'class' => $class ) ),
  $list_items_markup
);
```

I'm not opposed :)

Reply:
yes see #26192

Reply:
So essentially it would receive an array-like
```
gutenberg_get_block_props( array(
	'style' => 'border-color: #f00',
	'class' => 'has-border'
) );
```

And return an array with new props that should be added as attributes?

I guess ideally blocks would never add attributes to the block top-level node and always use our function that merges the attributes.

We may also have a function gutenberg_get_block_props_output that returns the attributes HTML definion of the attributes "style="..." class="..." given an array of attributes.

I think a path like that would make sense.

Reply:
If we continue with the current approach gutenberg_block_supports_classes what is the plan for the supports that add styles? Is having a gutenberg_block_supports_styles function the next step?

Reply:
> If we continue with the current approach gutenberg_block_supports_classes what is the plan for the supports that add styles? Is having a gutenberg_block_supports_styles function the next step?

Right, I don't think that's a good solution.

The initial focus on classes came from wanting to see if restricting Block Supports to outputting only classes could be enough, but it's clear we should open up to more attributes.